### DISCUSSION TITLE
Raspberry Pi libcamera 2 streams

### DISCUSSION THREAD

**Discussion by bhaal2** [bluenviron/mediamtx#1934] [Questions]
Hi there,
First up, I haven't touched mediamtx at all just yet, but been looking through the discussions and googling, I know libcamera supports 2 video streams as well as a stills stream from the same camera at the same time.  Just wondering if mediamtx has the ability to access both video streams, that way I can serve up a hires and lowres stream for cctv usage.  I have seen the discussions regarding using ffmpeg to create a 2nd stream, but this produces significant overhead that I'd like to avoid.

Thanks :)

**Comment by aler9** [comment]
Hello, the server already exploits the libcamera ability to generate 2 streams at once, each with a different resolution, the problem is that this ability is used to implement `rpiCameraMode`, in the same way as libcamera-vid does with `--mode`: one stream is used to pull the video, the other is used to set the sensor mode:

https://github.com/bluenviron/mediamtx/blob/20a3b07d0a218e74addaf5a1b5e70ca13a389727/internal/rpicamera/exe/camera.cpp#L154-L169

i don't know how this can be combined with a hi-res/low-res stream feature, since a stream is already occupied by the sensor mode, and we don't know in advance if the raspberry can encode both streams at once.

**Reply by bhaal2** [reply]
Thanks for getting back to me.  I feared this might be the case.  And I assume you mean there is only 1 channel in the h264 encoder anyway?

**Reply by aler9** [reply]
There are people that have been able to encode two streams at once with the hardware encoder by using FFmpeg and MediaMTX concurrently, eventually you have to decrease the resolution of one stream, but it can be done.

**Reply by bhaal2** [reply]
Ah I was trying this, but on a B+ a 1296x972 and a 640 x 480 ... ffmpeg was doing the 640x480, but it didn't work so well.  I have a bunch of A+, B+ and original Zeros running as CCTV cameras, I've been migrating them to mediamtx after using gstreamer stuff with the rpicam module Thaytan wrote years ago..  1 stream will have to do, at least by setting the mode to full res and then streaming at the larger of the above 2 resolutions has ended in a decent quality stream.  However on the A+ I've really had to fiddle with the GPU memory and the cma value, as well as stopping everything that's not needed from running in order to get the higher mode set...  Fun times.

**Comment by dragulesq** [comment]
@bhaal2 take a look at this [example](https://github.com/raspberrypi/picamera2/issues/713#issuecomment-1734136274)

You can change the `FileOutput` to be a UDP socket like in this [example](https://github.com/raspberrypi/picamera2/blob/main/examples/capture_stream_udp.py)

**Reply by dragulesq** [reply]
@aler9 given the example above, does it  mean that dual streams for the rpiCamera could be supported in mediamtx?
The example works for me on a pi zero w with cam v2 and v3

**Reply by bhaal2** [reply]
I'll take a look at that later this weekend, thank you for finding this!  Looks promising. @aler9 do we have to set the mode and therefore occupy the 2nd channel?  Can the mode not be assumed from the resolution being requested?

**Reply by dragulesq** [reply]
any luck with this?

[a comment by an automated account is not shown]
